state/protocol/badger/mutator.go

```diff
@@ -413,114 +416,137 @@ func (m *Mutator) sealExtend(candidate *flow.Block) (*flow.Seal, error) {
-	sealed, err := m.state.headers.ByBlockID(last.BlockID)
-	if err != nil {
-		return nil, fmt.Errorf("could not retrieve sealed block (%x): %w", last.BlockID, err)
-	}
-
-	var finalizedHeight uint64
-	err = m.state.db.View(operation.RetrieveFinalizedHeight(&finalizedHeight))
-	if err != nil {
-		return nil, fmt.Errorf("could not retrieve finalized height: %w", err)
+	// map each seal to the block it is sealing for easy lookup; we will need to
+	// successfully connect _all_ of these seals to the last sealed block for
+	// the payload to be valid
+	byBlock := make(map[flow.Identifier]*flow.Seal)
+	for _, seal := range payload.Seals {
+		byBlock[seal.BlockID] = seal
 	}
-	var finalID flow.Identifier
-	err = m.state.db.View(operation.LookupBlockHeight(finalizedHeight, &finalID))
-	if err != nil {
-		return nil, fmt.Errorf("could not lookup finalized block: %w", err)
+	if len(payload.Seals) != len(byBlock) {
+		return nil, state.NewInvalidExtensionErrorf("multiple seals for the same block")
 	}
 
-	// we now go from last sealed height plus one to finalized height and check
-	// if we have the seal for each of them step by step; often we will not even
-	// enter this loop, because last sealed height is higher than finalized
-	for height := sealed.Height + 1; height <= finalizedHeight; height++ {
-		// as we are iterating the finalized blocks, if there is all the seals
-		// have been used to seal the finalized blocks, and there is no more seal left,
-		// we could exit earlier with the last seal
-		if len(byBlock) == 0 {
-			return last, nil
-		}
-		header, err := m.state.headers.ByHeight(height)
-		if err != nil {
-			return nil, fmt.Errorf("could not get block for sealed height (%d): %w", height, err)
-		}
-		blockID := header.ID()
-		next, found := byBlock[blockID]
-		if !found {
-			return nil, state.NewInvalidExtensionErrorf("chain of seals broken for finalized (missing: %x)", blockID)
-		}
-		delete(byBlock, blockID)
-		last = next
-	}
-	// In case no seals are left, we skip the remaining part:
-	if len(byBlock) == 0 {
-		return last, nil
-	}
-	// Once we have filled in seals for all finalized blocks we need to check
-	// the non-finalized blocks backwards; collect all of them, from direct
-	// parent to just before finalized, and see if we can use up the rest of the
-	// seals. We need to stop collecting ancestors either when reaching the
-	// finalized state, or when reaching the last sealed block.
+	// incorporatedResults collects the _first_ appearance of unsealed execution
+	// results on the fork, along with the ID of the block in which they are
+	// incorporated.
+	incorporatedResults := make(map[flow.Identifier]*flow.IncorporatedResult)
+
+	// collect IDs of blocks on the fork (from parent to last sealed)
+	var blockIDs []flow.Identifier
+
+	// loop through the fork backwards up to last sealed block and collect
+	// IncorporatedResults as well as the IDs of blocks visited
+	sealedID := last.BlockID
 	ancestorID := header.ParentID
```

Comment:
Is there any reason we don't allow including an `ExecutionReceipt` and a `Seal` for the same block in the same payload? Handling this case would remove the extra blocks needed in `EpochBuilder` as well.

Reply:
> Is there any reason we don't allow including an ExecutionReceipt and a Seal

Yep there is a reason (this answer is slightly simplified; can go into more detail in a meeting):
* For BFT, the verifier assignment must be unpredictable at the time when the Exec Receipt is broadcast by the EN
* We guarantee this by:
   - incorporating Receipts into the block payload
   - using the _earliest_ occurrence of a particular execution result along a fork to determine the verifier assignment
   - specifically, we use the source of randomness from the block that incorporates a result for the first time in the fork 
* Therefore, in order to seal the result the following sequence of events must have occurred:
   1.  A receipt for the respective result was incorporated into the fork
   2.  A child of the block including the result must have been published (as only then the source of randomness is available)
   3. The verifiers approved the result 

In summary, the shortest possible latency to seal the result for a block `B0` is **3 additional child blocks**
```
 ... <- B0 <- B1{Exec. Receipt for B0: committing to result R0} <- B2{source of randomness for B1: allows computation of verifier's assignment} <- B3{seal for result R0}
```

Reply:
This is interesting. I hadn't thought about the need for the source of randomness block, and the test fixtures don't have it yet. This will further complicate the EpochBuilder and other fixtures. I am tempted to address this in phase 3, as changing the source of randomness is part of the scope of that phase.
@AlexHentschel @zhangchiqing

Reply:
It would actually be more precise to say that a payload cannot contain a `Seal` and the **specific `ExecutionReceipt` it is sealing**? It is possible for the payload to contain another receipt for the same block, because we store all receipts for accountability.

So if R(A) and R(A)' are two different receipts for the same block (A), we can have a situation like this:
```
  A         B             C
+---+    +------+    +---------+
|   | -> | R(A) | -> | S(R(A)) | 
+---+    +------+    | R(A)'   |
                     +---------+
```
Block C contains a seal for block A based on the first receipt committing to this block (R(A) in block B)

Block C also contains another receipt for block A.